modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,317 @@
+    return PAM_USER_UNKNOWN;
+  }
+
+  /* Get the password entry */
+  pwd = pam_modutil_getpwnam(pamh, user);
+  if (pwd == NULL) {
+    return PAM_CRED_INSUFFICIENT;
+  }
+
+  /* Check if we should not set quotas for user */
+  if ((pwd->pw_uid < param.start_uid) ||
+      ((param.end_uid >= param.start_uid) && (param.start_uid != 0) &&
+       (pwd->pw_uid > param.end_uid)))
+    return PAM_SUCCESS;
+
+  /* Find out what device the filesystem is hosted on */
+  if ((fp = setmntent("/etc/mtab", "r")) == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Unable to open /etc/mtab");
+    return PAM_PERM_DENIED;
+  }
+
+  while ((mnt = getmntent(fp)) != NULL) {
+    if (param.fs == NULL) {
+      /* If fs is not specified use filesystem with homedir as default */
+      size_t mnt_len = strlen(mnt->mnt_dir);
+      const char *s;
+      if (mnt_len > match_size &&
+         (s = pam_str_skip_prefix_len(pwd->pw_dir, mnt->mnt_dir, mnt_len)) != NULL &&
+         (s[0] == '\0' || s[0] == '/')) {
+        free(mntdevice);
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
+          pam_syslog(pamh, LOG_CRIT,  "Memory allocation error");
+          endmntent(fp);
+          return PAM_PERM_DENIED;
+        }
+        match_size = mnt_len;
+      }
+    } else if ((strcmp(param.fs, mnt->mnt_dir) == 0) ||
+               (strcmp(param.fs, mnt->mnt_fsname) == 0)) {
+        free(mntdevice);
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
+          pam_syslog(pamh, LOG_CRIT,  "Memory allocation error");
+          endmntent(fp);
+          return PAM_PERM_DENIED;
+        }
+    }
+  }
+
+  endmntent(fp);
+
+  if (mntdevice == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Filesystem not found");
+    return PAM_PERM_DENIED;
+  }
+
+  /* Get limits */
+  if (quotactl(QCMD(Q_GETQUOTA, USRQUOTA), mntdevice, pwd->pw_uid,
+               (void *)&ndqblk) == -1) {
+    pam_syslog(pamh, LOG_ERR, "fail to get limits for user %s : %m",
+               pwd->pw_name);
+    free(mntdevice);
+    return PAM_PERM_DENIED;
+  }
+
+  if (param.debug == 1)
+    debug(pamh, &ndqblk);
+
+  /* Only overwrite if quotas aren't already set or if overwrite is set */
+  if ((ndqblk.dqb_bsoftlimit == 0 && ndqblk.dqb_bhardlimit == 0 &&
+       ndqblk.dqb_isoftlimit == 0 && ndqblk.dqb_ihardlimit == 0) ||
+      param.overwrite == 1) {
+
+    /* Parse new limits
+     * Also warn about setting hard and soft limits via separate calls
+     * to pam_setquota. While this usually works, it can create errors
+     * when the quota for a user is reduced and the old soft limit is
+     * higher than the new hard limit.
+     * Setting both softlimit and hardlimit at the same time avoids that
+     * problem.
+     */
+    ndqblk.dqb_valid = 0;
+    if (!_pam_parse_dqblk(argc, argv, &ndqblk)) {
+      pam_syslog(pamh, LOG_AUTH | LOG_WARNING,
+        "Do not set soft and hard limits separately, undefined behavior might occur!");
+    }
```

Comment:
Why do we want to allow setting a hard limit without a soft limit, and visa versa?
Is there any potential usefulness in adding support for that?
There must be a reason for enabling this kind of user interface fragility in a new module.

Reply:
While I could think of some possible edge cases to just set the hardlimit and leave the softlimit alone, I have to agree that I don't really see any useful application here.
I will change the code the ouput an error should only one of the limits exit in the configuration.